### PR TITLE
feat(framework): Tags Input

### PR DESCRIPTION
**Pull request by SebCodesTheWeb**
closed DEV-9250

https://github.com/mediatool/northlight/assets/90923099/31a926f4-2980-4653-a145-d8a673abfcfb

### CODE REVIEW

**Review by orestisrodriguez — DISMISSED** [review]
this is called TaggableTextInput/TokenizedTextInput actually

File: framework/lib/components/tags-input/tags-input.tsx [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+ * multiple tags. Follows styling of textarea and is resizable.
+ *
+ * @see {@link https://northlight.dev/reference/tags-input}
+ * @example (Example)
+ * ## Simple
+ * (?
+ * <TagsInput />
+ * ?)
+ *
+ * @example (Example)
```
Comment (olchyk98) [inline framework/lib/components/tags-input/tags-input.tsx:35]: When options are specified, disallow values that are outside of scope.
Optionally, provide a separate prop that would control this behaviour.

**Review by olchyk98 — CHANGES_REQUESTED** [review]
approved by mistake

File: framework/lib/components/tags-input/tags-input.tsx [hunk 2]
```diff
@@ -0,0 +1,170 @@
+import React, {
```
Comment (olchyk98) [inline framework/lib/components/tags-input/tags-input.tsx:1]: The component needs unit tests.

**Review by olchyk98 — CHANGES_REQUESTED** [review]

File: framework/lib/components/tags-input/tags-input.tsx (hunk 2 above)
Comment (SebCodesTheWeb) [reply framework/lib/components/tags-input/tags-input.tsx:1]: Lmao, it is a complement when Oles reviews your PR and the almost  only thing he comments on is to add unit tests.

File: framework/lib/components/tags-input/tags-input.tsx [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+  const CustomSelect = CreatableSelect as FunctionComponent<TagsInputProps<T>>
+
+  const handleChange = (values: MultiValue<T>, actionMeta: ActionMeta<T>) => {
+    setSelectedOptions(values)
+    onChange(values, actionMeta)
+  }
+
+  const isValidNewOption = (input: string, availableOptions: MultiValue<T>) => {
+    const optionAlreadyExists = availableOptions.some(
```
Comment (olchyk98) [inline framework/lib/components/tags-input/tags-input.tsx:68]: `R.any`

File: framework/lib/components/tags-input/tags-input.tsx [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+      (option) => option.value === input
+    )
+    return input !== '' && !optionAlreadyExists
```
Comment (olchyk98) [inline framework/lib/components/tags-input/tags-input.tsx:71]: `!isEmpty(input)` or `!input`

File: framework/lib/components/tags-input/tags-input.tsx [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+  }
+
+  const addNewOption = (newOption: T) => {
+    onChange(selectedOptions, { action: 'select-option', option: newOption })
+    setSelectedOptions((prev) => [ ...prev, newOption ])
```
Comment (olchyk98) [inline framework/lib/components/tags-input/tags-input.tsx:76]: `setSelectedOptions(R.append(newOption))`

File: framework/lib/components/tags-input/tags-input.tsx [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    setInputValue('')
+  }
+
+  const handleKeyDown = (event: KeyboardEvent<HTMLDivElement>) => {
+    if (
+      !isValidNewOption(inputValue, selectedOptions) &&
+      inputValue !== '' &&
+      event.key !== ' '
+    ) {
+      setInputValue('')
```
Comment (olchyk98) [inline framework/lib/components/tags-input/tags-input.tsx:104]: i see this a lot, put it in a separate function: `clearInput` or something. That way, if we ever want to change the idea of `clear`, we'll have to modify only one single place.

File: framework/lib/components/tags-input/tags-input.tsx [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+      label: newInput.slice(0, -1),
+    }
+    addNewOption(newOption as T)
+    setInputValue('')
+  }
+
+  const handleKeyDown = (event: KeyboardEvent<HTMLDivElement>) => {
+    if (
+      !isValidNewOption(inputValue, selectedOptions) &&
+      inputValue !== '' &&
```
Comment (olchyk98) [inline framework/lib/components/tags-input/tags-input.tsx:101]: `!isEmpty(inputValue)`

File: framework/lib/components/tags-input/tags-input.tsx [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+      const newOption: Option = { value: inputValue, label: inputValue }
+      addNewOption(newOption as T)
+      setInputValue('')
+      event.preventDefault()
+    }
+  }
+
+  const prevOptions = useRef<OptionsOrGroups<T, GroupBase<T>> | undefined>(
+    options
+  )
```
Comment (olchyk98) [inline framework/lib/components/tags-input/tags-input.tsx:118]: \n

File: framework/lib/components/tags-input/tags-input.tsx [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+  const renderedOptions = useMemo(() => {
```
Comment (olchyk98) [inline framework/lib/components/tags-input/tags-input.tsx:119]: why does it have to be a `ref`? Can't it be a state value?
```
useEffect
  -> if(!equals(...)) setPreviousOptions(options)
  -> [options]
```

File: framework/lib/components/tags-input/tags-input.tsx [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+    if (!equals(prevOptions.current, options)) {
+      prevOptions.current = options
+    }
+    return prevOptions.current
+  }, [ options ])
+
+  const handleFocus = () => {
+    setIsFocused(true)
+    if (options && options.length > 0) {
```
Comment (olchyk98) [inline framework/lib/components/tags-input/tags-input.tsx:128]: options?.length > 0

**Review by olchyk98 — CHANGES_REQUESTED** [review]

File: framework/lib/components/tags-input/tags-input.tsx (hunk 1 above)
Comment (olchyk98) [reply framework/lib/components/tags-input/tags-input.tsx:35]: maybe, but it still feels weird to have suggestions and allow any value to be saved at the same time.
discard this comment for now

**Comment by olchyk98** [comment]
Just some small syntax stuff. I'm still a bit unsure about the suggestions logic.
Though it's nice to have, it's may be confusing for user to be allowed to pass any value, even if there are suggestions available.

File: framework/lib/components/tags-input/tags-input.tsx [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+      newInput === '' ||
+      newInput === ',' ||
+      !newInput.endsWith(',') ||
+      action === 'input-blur'
+    ) {
+      return
+    }
+    const newOption: Option = {
+      value: newInput,
+      label: newInput.slice(0, -1),
```
[the remark this answers is not in the document]
Comment (SebCodesTheWeb) [reply framework/lib/components/tags-input/tags-input.tsx:92]: Why?

File: framework/lib/components/tags-input/tags-input.tsx [hunk 12]
```diff
@@ -125,7 +127,7 @@ export function TagsInput<T extends Option> ({
 
   const handleFocus = () => {
     setIsFocused(true)
-    if (options && options.length > 0) {
+    if (options && options?.length > 0) {
```
Comment (olchyk98) [inline framework/lib/components/tags-input/tags-input.tsx:130]: `if(options?.length) {...}`

**Review by olchyk98 — CHANGES_REQUESTED** [review]

File: framework/lib/components/tags-input/tags-input.tsx (hunk 12 above)
Comment (SebCodesTheWeb) [reply framework/lib/components/tags-input/tags-input.tsx:130]: This throws an error that options?.length might be undefined

Comment (olchyk98) [reply framework/lib/components/tags-input/tags-input.tsx:130]: Weird. Are you trying to compare it to 0? Just do the direct value:
Undefined and 0 is false.

**Comment by olchyk98** [comment]
@SebCodesTheWeb Ready to be merged, but you'll have to address Orestis' comments first :)

**Comment by olchyk98** [comment]
@SebCodesTheWeb Pushed stuff to wrong branch. Last force-pushed commit contains files for step-stack.

**Review by olchyk98 — APPROVED** [review]

**Pull request merged**